pygsti/protocols/gst.py

```diff
@@ -1869,6 +1914,182 @@ def _add_badfit_estimates(results, base_estimate_label, badfit_options,
+    using the model introduced by Tim and Stefan in the RCSGST paper.
+    TODO: docstring (update)
+
+    Parameters
+    ----------
+    model : Model
+        The model to add a wildcard budget to.
+
+    ds : DataSet
+        The data the model predictions are being compared with.
+
+    circuits_to_use : list
+        The circuits whose data are compared.
+
+    parameters : dict
+        Various parameters of the estimate at hand.
+
+    badfit_options : GSTBadFitOptions, optional
+        Options specifying what post-processing actions should be performed when
+        a fit is unsatisfactory.  Contains detailed parameters for wildcard budget
+        creation.
+
+    comm : mpi4py.MPI.Comm, optional
+        An MPI communicator used to run this computation in parallel.
+
+    mem_limit : int, optional
+        A rough per-processor memory limit in bytes.
+
+    verbosity : int, optional
+        Level of detail printed to stdout.
+
+    Returns
+    -------
+    PrimitiveOpsWildcardBudget
+    """
+    printer = _baseobjs.VerbosityPrinter.create_printer(verbosity, mdc_objfn.resource_alloc)
+    badfit_options = GSTBadFitOptions.cast(badfit_options)
+    model = mdc_objfn.model
+    ds = mdc_objfn.dataset
+    global_circuits_to_use = mdc_objfn.global_circuits
+
+    printer.log("******************* Adding Wildcard Budget **************************")
+    
+    #cache construction code.
+    # Extract model, dataset, objfn, etc.
+    # Note: must evaluate mdc_objfn *before* passing to wildcard fn init so internal probs are init
+    mdc_objfn.fn(model.to_vector())
+    
+    ## PYGSTI TRANSPLANT from pygsti.protocols.gst._compute_wildcard_budget
+    # Compute the various thresholds
+    ds_dof = ds.degrees_of_freedom(global_circuits_to_use)
+    nparams = model.num_modeltest_params  # just use total number of params
+    percentile = 0.025
+    nboxes = len(global_circuits_to_use)
+
+    two_dlogl_threshold = _chi2.ppf(1 - percentile, max(ds_dof - nparams, 1))
+    redbox_threshold  = _chi2.ppf(1 - percentile / nboxes, 1)
+    
+    # Not in cache, recompute
+    ddists = _compute_ddists(estimate)
+    
+    critical_percircuit_budgets = _opt.wildcardopt._get_critical_circuit_budgets(mdc_objfn, redbox_threshold)
+
+    #In Stefan's original code these were all passed in as a dictionary, instead pass these in as arguments.
+    
+    alpha = _bisect_alpha(0.1, 1e-3, mdc_objfn, two_dlogl_threshold,
+                          redbox_threshold, ddists, critical_percircuit_budgets)
+   
+    return _ddist_wildcard_model(alpha, ddists)
+    
+def _bisect_alpha(guess, tol, mdc_objfn, two_dlogl_threshold, redbox_threshold, 
+                  ddists, critical_percircuit_budgets):
+    
+    left = None
+    right = None
+    
+    while left is None or right is None:
+        print(f'Searching for interval [{left}, {right}] with guess {guess}')
```

Comment:
We probably want to make this just debugger/info logging rather than print all the time.

Reply:
Good point, just pushed a commit implementing that.